### PR TITLE
With temp libpaths action

### PR DESCRIPTION
**Pull request by jimhester**
This PR does two things
- Adds a `rename` argument to `with_()`, which if `FALSE` will not try to rename the first argument to `new`. This is useful if the set function does not take a value to set (such as set_temp_libpaths())`.
- Adds a `action` argument to `set_temp_libpaths()` so you can control if you want to prefix, suffix or replace `.libPaths()`.

@krlmlr PTAL when you can, particularly would like feedback on the `rename` API.

### CODE REVIEW

File: R/libpaths.R [hunk 1]
```diff
@@ -35,8 +35,9 @@ with_libpaths <- with_(set_libpaths, .libPaths)
 #'
 #' Temporarily prepend a new temporary directory to the library paths.
 #'
-#' @template with
```
Comment (krlmlr) [inline R/libpaths.R]: Did you remove this by accident?

File: R/with_.R [hunk 2]
```diff
@@ -60,9 +68,17 @@ with_ <- function(set, reset = set, envir = parent.frame()) {
   set_call <- as.call(c(substitute(set), called_fmls))
 
   fun <- eval(bquote(function(args) {
-    old <- .(set_call)
```
Comment (krlmlr) [inline R/with_.R]: This change looks unrelated to me. I think I see what it does, but I'm not sure I like the way it's triggered by a missing arg.

File: R/libpaths.R [hunk 3]
```diff
@@ -12,10 +12,10 @@ set_libpaths <- function(paths, action = "replace") {
   invisible(old)
 }
 
-set_temp_libpath <- function() {
+set_temp_libpath <- function(action = "prefix") {
```
Comment (krlmlr) [inline R/libpaths.R]: How about `function(paths = tempfile("temp_libpath"), action = "prefix")`? This would allow postponing the "rename" change.

File: R/with_.R [hunk 4]
```diff
@@ -38,7 +42,7 @@
 #' }
 #' with_(set_global_state, reset_global_state)
 #' @export
-with_ <- function(set, reset = set, envir = parent.frame()) {
+with_ <- function(set, reset = set, envir = parent.frame(), rename = TRUE) {
```
Comment (krlmlr) [inline R/with_.R]: The "rename" arg looks like a logical extension if we need to support setters without first argument. But is it really necessary to support such setters?

File: R/with_.R (hunk 2 above)
Comment (jimhester) [inline R/with_.R]: You are correct, clearly forgot to stash my changes before doing this.

File: R/libpaths.R (hunk 1 above)
Comment (jimhester) [inline R/libpaths.R]: Yes thanks for the catch

File: R/libpaths.R (hunk 3 above)
Comment (jimhester) [inline R/libpaths.R]: That is true, probably worthwhile in this case as it also allows passing and alternative file if desired.

File: R/with_.R (hunk 4 above)
Comment (jimhester) [inline R/with_.R]: Probably not nessesary, we can table it for now.

File: R/libpaths.R (hunk 3 above)
Comment (krlmlr) [inline R/libpaths.R]: Or perhaps even have this as a default argument to with_libpaths(), and deprecate with_temp_libpaths()?

**Comment by jimhester** [comment]
https://github.com/jimhester/withr/commit/a231a3dcef5f80fe83506ef81c5a5ff8d8b5f442 uses your suggested approach with a default argument to `set_temp_libpaths()`.

`with_()` actually was not preserving the default argument if one existed, so I tweaked that to use the default argument if it exists.
